File: CoreAPI/Services/PatchService.cs
using Core.Exceptions;
using Core.Extensions;
using Core.Models;
using Core.ViewModels;
using CoreAPI.Services.Interfaces;
using CoreAPI.Services.Sql;
using CoreAPI.BgService;
using Microsoft.Extensions.Caching.Distributed;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Newtonsoft.Json;
using System.Text.RegularExpressions;

namespace CoreAPI.Services;

public class PatchService : IPatchService
{
    private readonly ISqlProvider _sql;
    private readonly IDistributedCache _cache;
    private readonly IConfiguration _configuration;
    private readonly IServiceProvider _serviceProvider;
    private readonly ILogger<PatchService> _logger;

    // User context - these should be set from the current request
    public string UserId { get; set; }
    public string TenantCode { get; set; }
    public string Env { get; set; }
    public List<string> RoleIds { get; set; } = new();
    public List<string> CenterIds { get; set; } = new();

    public PatchService(
        ISqlProvider sql,
        IDistributedCache cache,
        IConfiguration configuration,
        IServiceProvider serviceProvider,
        ILogger<PatchService> logger)
    {
        _sql = sql;
        _cache = cache;
        _configuration = configuration;
        _serviceProvider = serviceProvider;
        _logger = logger;

        _sql.TenantCode = TenantCode;
        _sql.Env = Env;
    }

    public void SetUserContext(string userId, string tenantCode, string env, List<string> roleIds)
    {
        UserId = userId;
        TenantCode = tenantCode;
        Env = env;
        RoleIds = roleIds;

        _sql.TenantCode = tenantCode;
        _sql.Env = env;
        _sql.UserId = userId;
        _sql.SystemFields = new List<string> { "id", "insertedby", "inserteddate", "updatedby", "updateddate" };
        if (tenantCode?.ToLower() != "system") _sql.SystemFields.Add("tenantcode");
    }

    public async Task<int> SavePatch(PatchVM vm)
    {
        vm.CachedDataConn ??= _sql.GetConnStrFromKey(vm.DataConn, vm.TenantCode, vm.Env);
        vm.CachedMetaConn ??= _sql.GetConnStrFromKey(vm.MetaConn, vm.TenantCode, vm.Env);
        var canWrite = await HasWritePermission(vm);
        if (!canWrite) throw new ApiException($"Unauthorized to write on \"{vm.Table}\"")
        {
            StatusCode = Core.Enums.HttpStatusCode.Unauthorized
        };
        var cmd = _sql.GetCreateOrUpdateCmd(vm);
        if (string.IsNullOrWhiteSpace(cmd)) return 0;
        var result = await _sql.RunSqlCmd(vm.CachedDataConn, cmd);
        return result;
    }

    public async Task<int> UpdatePatch(PatchVM vm)
    {
        vm.CachedDataConn ??= _sql.GetConnStrFromKey(vm.DataConn, vm.TenantCode, vm.Env);
        vm.CachedMetaConn ??= _sql.GetConnStrFromKey(vm.MetaConn, vm.TenantCode, vm.Env);
        var canWrite = await HasWritePermission(vm);
        if (!canWrite) throw new ApiException($"Unauthorized to write on \"{vm.Table}\"")
        {
            StatusCode = Core.Enums.HttpStatusCode.Unauthorized
        };
        var cmd = _sql.GetUpdateCmd(vm);
        if (string.IsNullOrWhiteSpace(cmd)) return 0;
        var result = await _sql.RunSqlCmd(vm.CachedDataConn, cmd);
        return result;
    }

    public async Task<SqlResult> SavePatch2(PatchVM vm)
    {
        // Simplified implementation - the full one is complex
        var id = vm.Changes.FirstOrDefault(x => x.Field == "Id")?.Value;
        if (string.IsNullOrEmpty(id))
        {
            return new SqlResult { status = 400, message = "Id is required" };
        }

        vm.CachedDataConn ??= _sql.GetConnStrFromKey(vm.DataConn, vm.TenantCode, vm.Env);

        if (id.StartsWith("-"))
        {
            // Insert new
            var (dup, mess, currentEntity) = await CheckDuplicate(vm);
            if (dup)
            {
                return new SqlResult
                {
                    updatedItem = null,
                    status = 409,
                    message = FormatEntityMessage(mess, currentEntity)
                };
            }
            id = id.Substring(1);
            AddDefaultFields(vm.Changes, new List<PatchDetail>
            {
                new PatchDetail { Field = "InsertedDate", Value = DateTime.Now.ToString("yyyy-MM-ddTHH:mm:ss") },
                new PatchDetail { Field = "InsertedBy", Value = UserId },
                new PatchDetail { Field = "UpdatedDate", Value = null },
                new PatchDetail { Field = "UpdatedBy", Value = null },
                new PatchDetail { Field = "Active", Value = "1" }
            });

            var cells = vm.Changes.Select(x => x.Field).ToList();
            var update = vm.Changes.Select(x => $"@{id + x.Field.ToLower()}");
            var cmd = $"INSERT into \"{vm.Table}\"(\"{cells.Combine("\",\"")}\") values({update.Combine()})";

            var result = await _sql.RunSqlCmd(vm.CachedDataConn, cmd);
            if (result > 0)
            {
                var entity = await _sql.ReadDataSet($"SELECT * FROM \"{vm.Table}\" WHERE \"Id\" = '{id}'");
                return new SqlResult
                {
                    updatedItem = entity[0],
                    status = 200,
                    message = "create successful"
                };
            }
        }
        else
        {
            // Update existing
            var (dup, mess, currentEntity) = await CheckDuplicate(vm, true);
            if (dup)
            {
                var entity = await _sql.ReadDataSet($"SELECT * FROM \"{vm.Table}\" WHERE \"Id\" = '{id}'");
                return new SqlResult
                {
                    updatedItem = entity[0],
                    status = 409,
                    message = FormatEntityMessage(mess, currentEntity)
                };
            }

            AddDefaultFields(vm.Changes, new List<PatchDetail>
            {
                new PatchDetail { Field = "UpdatedDate", Value = DateTime.Now.ToString("yyyy-MM-ddTHH:mm:ss") },
                new PatchDetail { Field = "UpdatedBy", Value = UserId }
            });

            var updates = vm.Changes.Where(x => x.Field != "Id").ToList();
            var update = updates.Select(x => $"\"{x.Field}\" = @{id + x.Field.ToLower()}");
            var cmd = $"UPDATE \"{vm.Table}\" SET {update.Combine()} WHERE \"Id\" = '{id}'";

            var result = await _sql.RunSqlCmd(vm.CachedDataConn, cmd);
            if (result > 0)
            {
                var entity = await _sql.ReadDataSet($"SELECT * FROM \"{vm.Table}\" WHERE \"Id\" = '{id}'");
                return new SqlResult
                {
                    updatedItem = entity[0],
                    status = 200,
                    message = "update successful"
                };
            }
        }

        return new SqlResult { status = 500, message = "Operation failed" };
    }

    public async Task<SqlResult> SavePatchs2(List<PatchVM> vms)
    {
        var results = new List<SqlResult>();
        foreach (var vm in vms)
        {
            var result = await SavePatch2(vm);
            results.Add(result);
            if (result.status != 200) break;
        }

        var successCount = results.Count(r => r.status == 200);
        return new SqlResult
        {
            status = successCount == vms.Count ? 200 : 207,
            message = $"{successCount}/{vms.Count} operations successful"
        };
    }

    public async Task<int> SavePatches(PatchVM[] patches)
    {
        if (patches == null || patches.Length == 0)
            throw new ArgumentException("patches is null or empty");

        patches = patches.Where(x => x.Id != null).ToArray();
        if (patches.Length == 0) return 0;

        patches[0].CachedDataConn ??= _sql.GetConnStrFromKey(patches[0].DataConn);
        patches[0].CachedMetaConn ??= _sql.GetConnStrFromKey(patches[0].MetaConn);

        var tables = patches.Select(x => x.Table);
        string rightQuery = $"select * from \"FeaturePolicy\" " +
            $"where \"Active\" = true and (\"CanWrite\" = true or \"CanWriteAll\" = true) and \"EntityName\" in ({tables.CombineStrings()}) and \"RoleId\" in ({RoleIds.CombineStrings()})";

        var permissions = await _sql.ReadDsAsArr<Core.Models.FeaturePolicy>(rightQuery, patches[0].CachedMetaConn);
        permissions = permissions.DistinctBy(x => x.TableName).ToArray();

        var lackPerTables = patches.Select(x => x.Table).Except(permissions.Select(x => x.TableName)).ToArray();
        if (lackPerTables.Length > 0)
        {
            throw new ApiException($"All table must have write permission {lackPerTables.CombineStrings()}")
            {
                StatusCode = Core.Enums.HttpStatusCode.Unauthorized
            };
        }

        var sql = patches.Select(_sql.GetCreateOrUpdateCmd).Where(x => x != null).Combine(";\n");
        var result = await _sql.RunSqlCmd(patches[0].CachedDataConn, sql);
        return result;
    }

    public async Task<bool> HardDelete(PatchVM vm)
    {
        if (vm.Delete == null || vm.Delete.Count == 0) return true;

        if (string.IsNullOrWhiteSpace(vm.ComId))
        {
            var sql = vm.Delete.Select(x => $"delete from \"{x.Table}\" where \"Id\" in ({x.Ids.CombineStrings()})");
            try
            {
                await _sql.RunSqlCmd(null, sql.Combine(";"));
                return true;
            }
            catch
            {
                return false;
            }
        }
        else
        {
            var query = $"select * from \"Component\" where \"Id\" = '{vm.ComId}' limit 1";
            var com = await _sql.ReadDsAs<Core.Models.Component>(query);
            var data = JsonConvert.DeserializeObject<SqlQuery>(com.Query);

            Dictionary<string, object> dictionary = new Dictionary<string, object>
            {
                { "EntityIds", vm.Delete.SelectMany(x => x.Ids).CombineStrings() },
                { "NewId", vm.NewId }
            };

            if (!string.IsNullOrWhiteSpace(data?.update))
            {
                var qr = FormatEntityMessage(data.update, dictionary);
                var deletequery = qr + ";" + vm.Delete.Select(x => $"delete from \"{x.Table}\" where \"Id\" in ({x.Ids.CombineStrings()})").Combine(";");
                try
                {
                    await _sql.RunSqlCmd(null, deletequery);
                    return true;
                }
                catch
                {
                    return false;
                }
            }
            else
            {
                var sql = vm.Delete.Select(x => $"delete from \"{x.Table}\" where \"Id\" in ({x.Ids.CombineStrings()})");
                try
                {
                    await _sql.RunSqlCmd(null, sql.Combine(";"));
                    return true;
                }
                catch
                {
                    return false;
                }
            }
        }
    }

    public async Task<string[]> DeactivateAsync(SqlViewModel vm)
    {
        vm.CachedDataConn ??= _sql.GetConnStrFromKey(vm.DataConn ?? "default");
        var allRights = await GetEntityPerm(vm.Table, null, vm.CachedDataConn);
        var canDeactivateAll = allRights.Any(x => x.CanDeactivateAll);
        var canDeactivateSelf = allRights.Any(x => x.CanDeactivate);

        var query = $"select * from \"{vm.Table}\" where \"Id\" in ({vm.Id.CombineStrings()})";
        var ds = await _sql.ReadDataSet(query, vm.CachedDataConn);
        var rows = ds.Length > 0 ? ds[0] : null;

        if (rows == null) return Array.Empty<string>();

        var unauthorized = new List<string>();
        foreach (var row in rows)
        {
            var isOwner = IsOwner(row, UserId, RoleIds);
            if (!canDeactivateAll && !(canDeactivateSelf && isOwner))
            {
                unauthorized.Add(row.GetValueOrDefault("Id")?.ToString() ?? "");
            }
        }

        if (unauthorized.Count > 0)
        {
            return unauthorized.ToArray();
        }

        var activeField = rows[0].ContainsKey("Active") ? "Active" : "IsActive";
        var updateQuery = $"UPDATE \"{vm.Table}\" SET \"{activeField}\" = false WHERE \"Id\" in ({vm.Id.CombineStrings()})";
        await _sql.RunSqlCmd(vm.CachedDataConn, updateQuery);

        return Array.Empty<string>();
    }

    private async Task<bool> HasWritePermission(PatchVM vm)
    {
        if (vm.ByPassPerm) return true;

        bool writePerm = false;
        var allRights = vm.ByPassPerm ? Array.Empty<Core.Models.FeaturePolicy>() : await GetEntityPerm(vm.Table, recordId: null, vm.CachedMetaConn);
        var idField = vm.Changes.FirstOrDefault(x => x.Field == "Id");
        var oldId = idField?.OldVal;

        if (oldId == null)
        {
            writePerm = allRights.Any(x => x.CanWriteAll);
        }
        else
        {
            var origin = $"select t.* from \"{vm.Table}\" as t where t.\"Id\" = '{oldId}'";
            var ds = await _sql.ReadDataSet(origin, vm.CachedDataConn);
            var originRow = ds.Length > 0 && ds[0].Length > 0 ? ds[0][0] : null;
            var isOwner = IsOwner(originRow, UserId, RoleIds);
            writePerm = isOwner || allRights.Any(x => x.CanWriteAll);
        }
        return writePerm;
    }

    private async Task<Core.Models.FeaturePolicy[]> GetEntityPerm(string entityName, string recordId, string connStr)
    {
        if (string.IsNullOrWhiteSpace(entityName)) return Array.Empty<Core.Models.FeaturePolicy>();
        if (RoleIds.Count == 0) return Array.Empty<Core.Models.FeaturePolicy>();

        var key = entityName + "_AllRights";
        var permissionByComCache = await GetStringAsync(key);

        Core.Models.FeaturePolicy[] permissions;
        if (!string.IsNullOrWhiteSpace(permissionByComCache))
        {
            permissions = JsonConvert.DeserializeObject<Core.Models.FeaturePolicy[]>(permissionByComCache);
        }
        else
        {
            var q = $"select * from \"FeaturePolicy\" " +
                $"where \"Active\" = true and \"EntityName\" = '{entityName}' " +
                $"and (\"RecordId\" = '{recordId}' or '{recordId}' = '') and \"RoleId\" in ({RoleIds.CombineStrings()})";

            permissions = await _sql.ReadDsAsArr<Core.Models.FeaturePolicy>(q, connStr);
            await SetStringAsync(key, JsonConvert.SerializeObject(permissions), TimeSpan.FromMinutes(5));
        }

        return permissions;
    }

    private async Task<(bool, string, Dictionary<string, object>)> CheckDuplicate(PatchVM patch, bool Update = false)
    {
        // Simplified - full implementation would check unique constraints
        return (false, "", new Dictionary<string, object>());
    }

    private string FormatEntityMessage(string template, Dictionary<string, object> data)
    {
        if (string.IsNullOrEmpty(template)) return "";

        var result = template;
        foreach (var kvp in data)
        {
            result = result.Replace("{" + kvp.Key + "}", kvp.Value?.ToString() ?? "");
        }
        return result;
    }

    private bool IsOwner(Dictionary<string, object> row, string userId, List<string> roleIds)
    {
        if (row == null) return false;

        var insertedBy = row.GetValueOrDefault("InsertedBy")?.ToString();
        return insertedBy == userId || roleIds.Contains("ADMIN");
    }

    public void AddDefaultFields(List<PatchDetail> changes, List<PatchDetail> defaultFields)
    {
        foreach (var field in defaultFields)
        {
            var existingField = changes.FirstOrDefault(change => change.Field == field.Field);
            if (existingField != null)
            {
                existingField.Value = field.Value;
            }
            else
            {
                changes.Add(field);
            }
        }
    }

    public async Task<Dictionary<string, object>[][]> GetTableColumns(string tableName)
    {
        var query = $@"
        SELECT c.COLUMN_NAME
        FROM INFORMATION_SCHEMA.COLUMNS c
        JOIN sys.columns sc ON c.COLUMN_NAME = sc.name
        AND OBJECT_ID(c.TABLE_SCHEMA + '.' + c.TABLE_NAME) = sc.object_id
        WHERE c.TABLE_NAME = '{tableName}'
        AND sc.is_computed = 0";

        return await _sql.ReadDataSet(query);
    }

    // Cache helpers
    private async Task<string> GetStringAsync(string key)
    {
        return await _cache.GetStringAsync(key);
    }

    private async Task SetStringAsync(string key, string value, TimeSpan? expiry = null)
    {
        if (expiry.HasValue)
            await _cache.SetStringAsync(key, value, new DistributedCacheEntryOptions { AbsoluteExpirationRelativeToNow = expiry });
        else
            await _cache.SetStringAsync(key, value);
    }
}

File: svc/src/services/patchService.ts
/**
 * Patch Service - CRUD Operations
 * Handles create, update, delete, and deactivate operations for database records
 * Migration from CoreAPI PatchService.cs
 */

import { query, execute, insert, update, remove } from "../database/postgresClient.ts";
import type { PatchVM, PatchDetail, FeaturePolicy, SqlResult, SqlViewModel, UserContext } from "../types/interfaces.ts";

/**
 * PatchService - Handles CRUD operations with permission checking
 */
export class PatchService {
  private userContext: UserContext | null = null;
  private cache: Map<string, { data: any; expiry: number }> = new Map();
  private readonly CACHE_TTL = 5 * 60 * 1000; // 5 minutes

  /**
   * Set the user context for the current request
   */
  setUserContext(context: UserContext): void {
    this.userContext = context;
  }

  /**
   * Get current user ID
   */
  private get userId(): string {
    return this.userContext?.userId || "";
  }

  /**
   * Get current tenant code
   */
  private get tenantCode(): string {
    return this.userContext?.tenantCode || "system";
  }

  /**
   * Get current environment
   */
  private get env(): string {
    return this.userContext?.env || "prod";
  }

  /**
   * Get current role IDs
   */
  private get roleIds(): string[] {
    return this.userContext?.roles || [];
  }

  /**
   * Cache helper - get value
   */
  private async getCache(key: string): Promise<any | null> {
    const cached = this.cache.get(key);
    if (cached && cached.expiry > Date.now()) {
      return cached.data;
    }
    return null;
  }

  /**
   * Cache helper - set value
   */
  private async setCache(key: string, value: any, ttl: number = this.CACHE_TTL): Promise<void> {
    this.cache.set(key, { data: value, expiry: Date.now() + ttl });
  }

  /**
   * SavePatch - Create or update a single record
   * @param patchVM - The patch view model containing table, changes, and metadata
   * @returns Number of affected rows
   */
  async SavePatch(patchVM: PatchVM): Promise<number> {
    // Check write permission
    const canWrite = await this.hasWritePermission(patchVM);
    if (!canWrite) {
      throw new Error(`Unauthorized to write on "${patchVM.table}"`);
    }

    const idField = patchVM.changes?.find((c: PatchDetail) => c.field.toLowerCase() === "id");
    const id = idField?.value || "";
    const isNew = !id || id.startsWith("-");

    if (isNew) {
      return await this.insertRecord(patchVM);
    } else {
      return await this.updateRecord(patchVM);
    }
  }

  /**
   * SavePatch2 - Alternative save with explicit ID handling
   * Returns SqlResult with status and data
   * @param patchVM - The patch view model
   * @returns SqlResult with operation status
   */
  async SavePatch2(patchVM: PatchVM): Promise<SqlResult> {
    const table = patchVM.table;
    if (!table) {
      return { message: "Table is required", status: 400 };
    }

    const idField = patchVM.changes?.find((c: PatchDetail) => c.field.toLowerCase() === "id");
    const id = idField?.value;

    if (!id) {
      return { message: "Id is required", status: 400 };
    }

    // Check write permission
    const canWrite = await this.hasWritePermission(patchVM);
    if (!canWrite) {
      return { message: `Unauthorized to write on "${table}"`, status: 401 };
    }

    if (id.startsWith("-")) {
      // Insert new record
      return await this.insertRecordWithResult(patchVM, id.substring(1));
    } else {
      // Update existing record
      return await this.updateRecordWithResult(patchVM, id);
    }
  }

  /**
   * SavePatchs - Batch save multiple records
   * @param patchVMs - Array of patch view models
   * @returns SqlResult with aggregated results
   */
  async SavePatchs(patchVMs: PatchVM[]): Promise<SqlResult> {
    if (!patchVMs || patchVMs.length === 0) {
      return { message: "No patches to save", status: 400 };
    }

    const results: SqlResult[] = [];
    let successCount = 0;

    for (const patchVM of patchVMs) {
      const result = await this.SavePatch2(patchVM);
      results.push(result);
      if (result.status === 200) {
        successCount++;
      } else {
        // Stop on first error or continue based on requirements
        break;
      }
    }

    return {
      message: `${successCount}/${patchVMs.length} operations successful`,
      status: successCount === patchVMs.length ? 200 : 207,
      data: results.map(r => r.updatedItem || []).flat(),
    };
  }

  /**
   * HardDelete - Permanently delete records
   * @param table - The table name
   * @param id - The record ID(s) to delete
   * @returns Boolean indicating success
   */
  async HardDelete(table: string, id: string | string[]): Promise<boolean> {
    if (!table) {
      throw new Error("Table name is required");
    }

    const ids = Array.isArray(id) ? id : [id];
    if (ids.length === 0) {
      return true;
    }

    // Check delete permission
    const canDelete = await this.hasDeletePermission(table, ids);
    if (!canDelete) {
      throw new Error(`Unauthorized to delete from "${table}"`);
    }

    try {
      const idsString = ids.map((i: string) => `'${i}'`).join(",");
      const sql = `DELETE FROM "${table}" WHERE "Id" IN (${idsString})`;
      await execute(sql);
      return true;
    } catch (error) {
      console.error("HardDelete error:", error);
      return false;
    }
  }

  /**
   * DeactivateAsync - Soft delete (set Active = false)
   * @param table - The table name
   * @param id - The record ID(s) to deactivate
   * @returns Array of unauthorized IDs (empty if all succeeded)
   */
  async DeactivateAsync(table: string, id: string | string[]): Promise<string[]> {
    if (!table) {
      throw new Error("Table name is required");
    }

    const ids = Array.isArray(id) ? id : [id];
    if (ids.length === 0) {
      return [];
    }

    // Check deactivate permission
    const permission = await this.getEntityPermission(table);
    const canDeactivateAll = permission.some((p: FeaturePolicy) => p.canDeleteAll);
    const canDeactivateSelf = permission.some((p: FeaturePolicy) => p.canDeactivate);

    // Get current records to check ownership
    const idsString = ids.map((i: string) => `'${i}'`).join(",");
    const selectSql = `SELECT * FROM "${table}" WHERE "Id" IN (${idsString})`;
    const records = await query(selectSql);

    if (!records || records.length === 0) {
      return [];
    }

    // Check authorization for each record
    const unauthorized: string[] = [];
    for (const record of records) {
      const isOwner = this.isOwner(record, this.userId, this.roleIds);
      if (!canDeactivateAll && !(canDeactivateSelf && isOwner)) {
        unauthorized.push(record.Id || record.id || "");
      }
    }

    if (unauthorized.length > 0) {
      return unauthorized;
    }

    // Perform soft delete
    const activeField = records[0].hasOwnProperty("Active") ? "Active" :
                        records[0].hasOwnProperty("IsActive") ? "IsActive" : "active";
    const updateSql = `UPDATE "${table}" SET "${activeField}" = false WHERE "Id" IN (${idsString})`;
    await execute(updateSql);

    return [];
  }

  // ============================================
  // Private Helper Methods
  // ============================================

  /**
   * Insert a new record
   */
  private async insertRecord(patchVM: PatchVM): Promise<number> {
    const table = patchVM.table;
    if (!table) throw new Error("Table is required");

    // Add default fields for new records
    this.addDefaultFields(patchVM.changes || [], [
      { field: "InsertedDate", value: new Date().toISOString() },
      { field: "InsertedBy", value: this.userId },
      { field: "UpdatedDate", value: null },
      { field: "UpdatedBy", value: null },
      { field: "Active", value: "true" },
    ]);

    const data = this.changesToObject(patchVM.changes || []);
    const result = await insert(table, data);
    return result?.length || 0;
  }

  /**
   * Insert record and return SqlResult
   */
  private async insertRecordWithResult(patchVM: PatchVM, id: string): Promise<SqlResult> {
    const table = patchVM.table as string;

    // Add default fields
    this.addDefaultFields(patchVM.changes || [], [
      { field: "Id", value: id },
      { field: "InsertedDate", value: new Date().toISOString() },
      { field: "InsertedBy", value: this.userId },
      { field: "UpdatedDate", value: null },
      { field: "UpdatedBy", value: null },
      { field: "Active", value: "true" },
    ]);

    const data = this.changesToObject(patchVM.changes || []);

    try {
      const result = await insert(table, data);
      if (result && result.length > 0) {
        // Fetch the inserted record
        const selectResult = await query(`SELECT * FROM "${table}" WHERE "Id" = '${id}'`);
        return {
          message: "create successful",
          status: 200,
          updatedItem: selectResult,
        };
      }
      return { message: "Insert failed", status: 500 };
    } catch (error: any) {
      console.error("Insert error:", error);
      return { message: error.message || "Insert failed", status: 500 };
    }
  }

  /**
   * Update an existing record
   */
  private async updateRecord(patchVM: PatchVM): Promise<number> {
    const table = patchVM.table as string;
    if (!table) throw new Error("Table is required");

    const idField = patchVM.changes?.find((c: PatchDetail) => c.field.toLowerCase() === "id");
    const id = idField?.value;
    if (!id) throw new Error("ID is required for update");

    // Add default fields for updates
    this.addDefaultFields(patchVM.changes || [], [
      { field: "UpdatedDate", value: new Date().toISOString() },
      { field: "UpdatedBy", value: this.userId },
    ]);

    const data = this.changesToObject(patchVM.changes || []);
    // Remove Id from update data
    delete data.Id;
    delete data.id;

    const result = await update(table, data, { Id: id });
    return result?.length || 0;
  }

  /**
   * Update record and return SqlResult
   */
  private async updateRecordWithResult(patchVM: PatchVM, id: string): Promise<SqlResult> {
    const table = patchVM.table as string;

    // Add default fields
    this.addDefaultFields(patchVM.changes || [], [
      { field: "UpdatedDate", value: new Date().toISOString() },
      { field: "UpdatedBy", value: this.userId },
    ]);

    const data = this.changesToObject(patchVM.changes || []);
    // Remove Id from update data
    delete data.Id;
    delete data.id;

    try {
      const result = await update(table, data, { Id: id });
      if (result && result.length > 0) {
        // Fetch the updated record
        const selectResult = await query(`SELECT * FROM "${table}" WHERE "Id" = '${id}'`);
        return {
          message: "update successful",
          status: 200,
          updatedItem: selectResult,
        };
      }
      return { message: "Update failed", status: 500 };
    } catch (error: any) {
      console.error("Update error:", error);
      return { message: error.message || "Update failed", status: 500 };
    }
  }

  /**
   * Check if user has write permission for the table
   */
  private async hasWritePermission(patchVM: PatchVM): Promise<boolean> {
    // Bypass permission check if flag is set
    if (patchVM.byPassPerm) return true;

    const table = patchVM.table;
    if (!table) return false;

    const idField = patchVM.changes?.find((c: PatchDetail) => c.field.toLowerCase() === "id");
    const oldId = idField?.oldVal;

    const permissions = await this.getEntityPermission(table);

    if (!oldId) {
      // New record - check CanWriteAll
      return permissions.some((p: FeaturePolicy) => p.canWriteAll);
    } else {
      // Existing record - check ownership or CanWriteAll
      const recordQuery = `SELECT * FROM "${table}" WHERE "Id" = '${oldId}'`;
      const records = await query(recordQuery);
      const record = records?.[0];
      const isOwner = this.isOwner(record, this.userId, this.roleIds);
      return isOwner || permissions.some((p: FeaturePolicy) => p.canWriteAll);
    }
  }

  /**
   * Check if user has delete permission for the table
   */
  private async hasDeletePermission(table: string, ids: string[]): Promise<boolean> {
    if (!table || ids.length === 0) return false;

    const permissions = await this.getEntityPermission(table);

    // Check CanDeleteAll
    if (permissions.some((p: FeaturePolicy) => p.canDeleteAll)) {
      return true;
    }

    // Check CanDelete with ownership
    const canDeleteSelf = permissions.some((p: FeaturePolicy) => p.canDelete);
    if (!canDeleteSelf) return false;

    // Check ownership for all records
    const idsString = ids.map((i: string) => `'${i}'`).join(",");
    const records = await query(`SELECT * FROM "${table}" WHERE "Id" IN (${idsString})`);

    for (const record of records) {
      if (!this.isOwner(record, this.userId, this.roleIds)) {
        return false;
      }
    }

    return true;
  }

  /**
   * Get entity permissions from FeaturePolicy table
   */
  private async getEntityPermission(entityName: string, recordId?: string): Promise<FeaturePolicy[]> {
    if (!entityName || this.roleIds.length === 0) {
      return [];
    }

    const cacheKey = `${entityName}_permissions`;
    const cached = await this.getCache(cacheKey);
    if (cached) {
      return cached;
    }

    const roleIdsString = this.roleIds.map((r: string) => `'${r}'`).join(",");
    const sql = `
      SELECT * FROM "FeaturePolicy"
      WHERE "Active" = true
      AND "EntityName" = '${entityName}'
      AND ("RecordId" = '${recordId || ''}' OR '${recordId || ''}' = '')
      AND "RoleId" IN (${roleIdsString})
    `;

    try {
      const permissions = await query(sql);
      await this.setCache(cacheKey, permissions);
      return permissions;
    } catch (error) {
      console.error("Error fetching permissions:", error);
      return [];
    }
  }

  /**
   * Check if user is owner of a record
   * Owner = InsertedBy == UserId or user has ADMIN role
   */
  private isOwner(record: any, userId: string, roleIds: string[]): boolean {
    if (!record) return false;

    // Check if user has ADMIN role
    if (roleIds.includes("ADMIN") || roleIds.includes("admin")) {
      return true;
    }

    const insertedBy = record.InsertedBy || record.insertedBy || record.InsertedBy_userId;
    return insertedBy === userId;
  }

  /**
   * Add default fields to changes array
   */
  private addDefaultFields(changes: PatchDetail[], defaultFields: { field: string; value: any }[]): void {
    for (const field of defaultFields) {
      const existing = changes.find((c: PatchDetail) => c.field.toLowerCase() === field.field.toLowerCase());
      if (existing) {
        existing.value = field.value;
      } else {
        changes.push({
          field: field.field,
          value: field.value,
        });
      }
    }
  }

  /**
   * Convert PatchDetail array to object for database insert/update
   */
  private changesToObject(changes: PatchDetail[]): Record<string, any> {
    const obj: Record<string, any> = {};
    for (const change of changes) {
      if (change.field.toLowerCase() === "id") continue; // Skip Id field for data object
      obj[change.field] = this.parseValue(change.value);
    }
    return obj;
  }

  /**
   * Parse value to appropriate type
   */
  private parseValue(value: any): any {
    if (value === null || value === undefined) {
      return null;
    }
    if (value === "true" || value === "false") {
      return value === "true";
    }
    // Try parsing as number
    const num = Number(value);
    if (!isNaN(num) && value !== "") {
      return num;
    }
    return value;
  }
}

// Export singleton instance
export const patchService = new PatchService();
